### PR TITLE
Create InterruptibleCharSequence

### PR DESCRIPTION
**Pull request by fbruton**

### CODE REVIEW

**Review by sambish5 — DISMISSED** [review]

**Review by cfkoehler — APPROVED** [review]
Clean and simple. I was able to integrate upstream and force an interrupt

File: src/test/java/emissary/core/InterruptibleCharSequenceTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            tryMatch(inputWrapped, regex);
+            end = System.currentTimeMillis();
+            logger.info("InterruptibleCharSequence {}", end - start);
+            interruptibleCharSequenceTotalTime += end - start;
+        }
+        float overheadPercentage = (float) interruptibleCharSequenceTotalTime / stringTotalTime;
+        logger.info("InterruptibleCharSequence overhead percentage: {}.", overheadPercentage);
+    }
+
+    public boolean tryMatch(CharSequence input, String regex) {
```
Comment (drivenflywheel) [inline src/test/java/emissary/core/InterruptibleCharSequenceTest.java:48]: (optional - could make this method static)
```suggestion
    public static boolean tryMatch(CharSequence input, String regex) {
```

File: src/test/java/emissary/core/InterruptibleCharSequenceTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            }
+            t.interrupt();
+            Object result = blockingQueue.take();
+
+            // Assert that InterruptedByTimeoutException was thrown
+            if (result instanceof InterruptedByTimeoutException) {
+                return;
+            } else {
+                fail("failed to interrupt InterruptibleCharSequence with given sleeping intervals");
+            }
```
Comment (drivenflywheel) [inline src/test/java/emissary/core/InterruptibleCharSequenceTest.java:98]: Simplify?
```suggestion
            assertTrue(result instanceof InterruptedByTimeoutException, "failed to interrupt InterruptibleCharSequence with given sleeping intervals");
```

File: src/main/java/emissary/core/InterruptibleCharSequence.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    public InterruptibleCharSequence(CharSequence inner) {
+        super();
+        this.inner = inner;
+    }
+
+    @Override
+    public char charAt(int index) {
+        if (Thread.currentThread().isInterrupted()) {
+            throw new InterruptedByTimeoutException("Interrupted reading CharSequence");
```
Comment (drivenflywheel) [inline src/main/java/emissary/core/InterruptibleCharSequence.java:14]: I'm not sure this new exception type is needed: we can observe that the current thread is interrupted, but can't definitely assert that it's due to a timeout.  I think a standard `InterruptedException` might be sufficient.

File: src/test/java/emissary/core/InterruptibleCharSequenceTest.java (hunk 2 above)
Comment (fbruton) [reply src/test/java/emissary/core/InterruptibleCharSequenceTest.java:98]: The while loop needs to exit or the next iteration will return false. I struggled with making this code more concise and open to other suggestions.

File: src/main/java/emissary/core/InterruptibleCharSequence.java (hunk 3 above)
Comment (fbruton) [reply src/main/java/emissary/core/InterruptibleCharSequence.java:14]: I believe that any exception thrown from within `charAt` needs to be an instance of a `RuntimeException`. `InterruptedException` only extends `Exception`, which would require that `charAt` declare an exception can be thrown.

File: src/test/java/emissary/core/InterruptibleCharSequenceTest.java (hunk 2 above)
Comment (drivenflywheel) [reply src/test/java/emissary/core/InterruptibleCharSequenceTest.java:98]: Maybe?  
```java
   assertTrue(... same stuff);
   return;
```

**Review by drivenflywheel — APPROVED** [review]

File: src/test/java/emissary/core/InterruptibleCharSequenceTest.java (hunk 2 above)
Comment (fbruton) [reply src/test/java/emissary/core/InterruptibleCharSequenceTest.java:98]: It's caused the PMD verify to fail the build. Didn't like `return` being the last statement in the loop.

**Review by sambish5 — APPROVED** [review]

**Pull request merged**